feat: rag quiz cherrypick [AI-1172]

gclomax opened the pull request:
## Description

This PR introduces and tests the **RAG Quiz Recommender** functionality, enhancing how Aila recommends starter and exit quizzes using multiple strategies. Key changes include:

- Implementation of a new quiz generator: `AilaRagQuizGenerator`
- Addition of override logic to prioritize `BasedOnRag` quizzes if available
- New method `createBestQuizOverride()` within `BaseFullQuizService` that enables tiered quiz fallback logic
- Integration of `override: true` in `fetchExperimentalPatches` for testing end-to-end behavior
- Elastic index updates for lesson-slug and quiz-question sources to reflect `2025-04-16`
- Enhanced test coverage including override paths and fallback behavior when no `basedOn` lesson is present
- Normalized fallback behavior in `ElasticLessonQuizLookup` to return empty arrays instead of throwing
- Logger update to scope debug output

## Issue(s)

Fixes AI-1172

## How to test

1. Go to https://oak-ai-lesson-assistant-lq5pc7xzh.vercel-preview.thenational.academy  
2. Create a new Aila lesson and proceed to quiz generation  
3. You should see a quiz generated based on the following logic:  
   - If a `BasedOnRagQuizGenerator` produces results → that quiz is used  
   - Else, it falls back to `AilaRagQuizGenerator` or `MLQuizGenerator` output  
4. The quiz will be selected based on the `return-first` reranker setting  
5. You can observe different behavior by switching `quizGenerators` and toggling `override` to `false` (see `AilaChat.ts` line 94)

## Screenshots

How it used to look (delete if n/a):  
{screenshots}

How it should now look:  
{screenshots}

## Checklist

- [x] Manually tested across browsers / devices  
- [x] Considered impact on accessibility  
- [x] Does this PR update a package with a breaking change

gclomax committed:
feat: Add AilaRagQuizGenerator for dynamic quiz generation

Introduces new quiz generator for creating quizzes based on AILA RAG relevant lessons. Includes:
- AilaRagQuizGenerator class with methods for generating starter and exit quizzes
- Corresponding test suite for quiz generation functionality
- Support for mapping quizzes from lesson plans and retrieving questions

gclomax committed:
fix: Handle missing quiz gracefully in ElasticLessonQuizLookup

Modify searchQuizByLessonSlug to return an empty array instead of throwing an error when no quiz is found for a given lesson slug. This conforms with other functionality for recommenders returning empty quizzes if none found

gclomax committed:
feat: Implement RAG quiz generator and update test suite

- Add AilaRagQuizGenerator implementation in AilaQuizGeneratorFactory
- Update CompositeFullQuizServiceBuilder test to include RAG quiz generator test case
- Rename existing test to clarify BasedOnRag quiz generator test

gclomax committed:
Chore: Remove commented-out lesson plan entries in AilaRagQuizGenerator test

gclomax committed:
chore: updating generator for testing

gclomax committed:
added more tests

gclomax committed:
changed debug scope for temp deployment

gclomax committed:
feat: Enhance BaseFullQuizService with quiz generation override strategy

- Add new method `createBestQuizOverride` to support fallback quiz generation
- Implement logic to prioritize BasedOnRagQuizGenerator when available
- Fallback to other quiz generators if BasedOnRagQuizGenerator fails to produce a quiz
- Maintain existing quiz selection and ranking mechanisms

gclomax committed:
refactor: Modify quiz lookup to return empty array instead of throwing error

- Update `getStarterQuiz` and `getExitQuiz` methods to return empty array when no quiz is found
- Adjust `hasStarterQuiz` and `hasExitQuiz` methods to check array length
- Update corresponding test cases to reflect new behavior
- Improve error handling and consistency in quiz lookup methods

gclomax committed:
feat: Add override functionality to FullQuizService

- Introduce `override` parameter in `createBestQuiz` method to control quiz generation behavior
- Implement tests to validate functionality with and without override enabled
- Ensure quizzes are generated correctly based on the override setting

gclomax committed:
feat: Introduce CircleTheoremLessonWithoutBasedOn for quiz generation

- Add `CircleTheoremLessonWithoutBasedOn` export to facilitate quiz generation without the `basedOn` property.
- Update tests in `CompositeFullQuizServiceBuilder` to validate quiz generation behavior with and without the `basedOn` lesson.
- Ensure proper handling of quiz generation when the `basedOn` lesson is absent.

gclomax committed:
Merge branch 'feat/maths-search-generator-cherrypick' into feat/rag-quiz-cherrypick

codeincontext commented on packages/aila/src/core/quiz/LessonSlugQuizMapping.ts:
```diff
@@ -56,7 +57,8 @@ export class ElasticLessonQuizLookup extends BaseLessonQuizLookup {
       if (!response.hits.hits[0]?._source) {
         log.error(`No ${quizType} found for lesson slug: ${lessonSlug}. Hit: `);
-        throw new Error(`No ${quizType} found for lesson slug: ${lessonSlug}`);
+        // throw new Error(`No ${quizType} found for lesson slug: ${lessonSlug}`);
```
Is this an expected error or an exceptional case? If it's an exception then maybe we can report it to Sentry?

gclomax replied:
The error is expected for the current quizzes we are using. This will change when we only use non-legacy quizzes.

codeincontext commented on packages/aila/src/core/quiz/fullservices/BaseFullQuizService.ts:
```diff
@@ -66,6 +89,83 @@ export abstract class BaseFullQuizService implements FullQuizService {
+
+    const quizArrays = await Promise.all(quizPromises);
+    const quizzes = quizArrays.flat();
+    if (!this.quizReranker.ratingSchema) {
+      throw new Error("Reranker rating schema is undefined");
+    }
```
We sometimes use the invariant pattern for this

```
invariant(this.quizReranker.ratingSchema, "Reranker rating schema is undefined")
```

codeincontext committed:
chore: don't force colour in CI tests (#629)

stefl committed:
feat: .cursorrules and CLAUDE.md (#593)

codeincontext committed:
fix: add lesson plan tracking store (#619)

Co-authored-by: Joe Baker <joe.baker@thenational.academy>

codeincontext committed:
test: align @playwright/test and playwright package versions (#634)

codeincontext committed:
chore: update avo to main (#637)

codeincontext committed:
chore: update avo to main (#637)

gclomax committed:
chore: pr-feedback, changing to invariant

gclomax committed:
chore: pr-feeback, removing assertions

mikeritson-oak committed:
Merge pull request #638 from oaknational/rc-20250326-1

build: release candidate

JBR90 committed:
feat: additional material package (#597)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Stef Lewandowski <stef@stef.io>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mantagen commented on packages/aila/src/core/quiz/fullservices/BaseFullQuizService.ts:
```diff
@@ -66,6 +89,85 @@ export abstract class BaseFullQuizService implements FullQuizService {
+          ailaRagRelevantLessons,
+        );
+      }
+      throw new Error(`Invalid quiz type: ${quizType as string}`);
```
is `as string` necessary here?

mantagen commented on packages/aila/src/core/quiz/fullservices/CompositeFullQuizServiceBuilder.test.ts:
```diff
@@ -54,9 +57,9 @@ const shouldSkipTests = process.env.TEST_QUIZZES === "false";
+    it("Should work with a BasedOnRag quiz generator", async () => {
       const mockRelevantLessons: AilaRagRelevantLesson[] = [
-        { lessonPlanId: "clna7k8j400egp4qxrqmjx0qo", title: "test-title-2" },
+        // { lessonPlanId: "clna7k8j400egp4qxrqmjx0qo", title: "test-title-2" },
```
why remove this case?

mantagen commented on packages/aila/src/core/quiz/fullservices/CompositeFullQuizServiceBuilder.test.ts:
```diff
@@ -87,46 +90,174 @@ const shouldSkipTests = process.env.TEST_QUIZZES === "false";
         quizSelector: "simple",
-        quizReranker: "schema-reranker",
-        quizGenerators: ["basedOnRag", "ml"],
+        quizReranker: "return-first",
```
I feel like git has got a bit confused and things you're changing parts but they're actually new tests! Makes it harder to review

mantagen commented on packages/aila/src/core/quiz/generators/AilaRagQuizGenerator.ts:
```diff
@@ -0,0 +1,48 @@
+    ailaRagRelevantLessons: AilaRagRelevantLesson[],
+    quizType: QuizPath = "/starterQuiz",
+  ): Promise<Quiz[]> {
+    // TODO: MG - This is a load of DB queries and may make it spiky.
```
Yep we might want a nice API to tap into for doing this?

mantagen reviewed (approved):
Code looks good, couple questions but nothing blocking.

Could you fill out the "how to test" bit with expected behaviour for @mikeritson-oak and maybe Ed?

semantic-release-bot committed:
build(release v1.31.0): See CHANGE_LOG.md

actions-user committed:
Merge remote-tracking branch 'origin/production'

gclomax committed:
chore: turning off ml

gclomax committed:
chore: changing order of generators

gclomax committed:
fix: Fix issue with quiz generation logic and add tests for override functionality

- Update `BaseFullQuizService` to improve quiz array checks and ensure proper handling of quiz generation.
- Modify quiz generator filtering to include `AilaRagQuizGenerator` in specific conditions.
- Add new tests in `CompositeFullQuizServiceBuilder` to validate quiz generation with and without the `basedOn` lesson, including override functionality.
- Ensure quizzes are generated correctly based on the override setting, enhancing overall quiz generation behavior.

gclomax committed:
chore: turning override on

mikeritson-oak committed:
Merge pull request #645 from oaknational/rc-20250402-1

build: release candidate

semantic-release-bot committed:
build(release v1.32.0): See CHANGE_LOG.md

JBR90 committed:
refactor: additional material package AI-944 (#642)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Stef Lewandowski <stef@stef.io>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mantagen committed:
feat: maths tex notation (#577)

mikeritson-oak committed:
Merge pull request #649 from oaknational/rc-20250410-1

build: release candidate

semantic-release-bot committed:
build(release v1.33.0): See CHANGE_LOG.md

codeincontext committed:
chore: don't log zod schema (#636)

codeincontext committed:
fix: don't throw on slack notification failure (#621)

codeincontext committed:
test: 3 playwright workers in dev (#635)

codeincontext committed:
chore: emit error objects to console from logger (#632)

gclomax committed:
chore: Update Elasticsearch index names for quiz generation

- Change index names in `LessonSlugQuizMapping`, `AilaRagQuizGenerator`, and `BaseQuizGenerator` to include date suffix "2025-04-16" for better versioning.
- Ensure consistent naming across all quiz-related components to improve maintainability and clarity.

gclomax committed:
chore: Update QuizQuestionTextOnlySource to allow raw JSON data

- Modify the `metadata` field in `QuizQuestionTextOnlySource` to include an optional `raw_json` property for better flexibility in handling raw JSON data.
- This change enhances the interface to accommodate additional data formats as needed.

gclomax committed:
chore: Update CustomSource and BaseQuizGenerator for improved quiz handling

- Enhance the `CustomSource` interface by adding new fields: `questionUid`, `lessonSlug`, `quizPatchType`, `isLegacy`, and `embedding` to better manage quiz data.
- Modify `BaseQuizGenerator` to reflect changes in the data structure, ensuring proper handling of the new fields and updating Elasticsearch queries accordingly.
- Update the mapping of `custom_id` to use `questionUid` from the source for consistency in quiz generation.

JBR90 committed:
feat: additional resource from partial aila lesson (#644)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Stef Lewandowski <stef@stef.io>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

tomwisecodes committed:
chore: remove useless logs (#651)

mikeritson-oak committed:
Merge pull request #653 from oaknational/rc-20250422-1

build: release candidate

semantic-release-bot committed:
build(release v1.34.0): See CHANGE_LOG.md

dependabot[bot] committed:
build(deps-dev): bump axios from 1.7.7 to 1.8.2 (#641)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

tomwisecodes committed:
feat: banner implementaion (#647)

tomwisecodes committed:
feat: add the hubspot contact record id to posthog (#650)

gclomax committed:
chore: pushing up changes to maths generator

mikeritson-oak committed:
Merge pull request #661 from oaknational/rc-20250429-1

build: release candidate

semantic-release-bot committed:
build(release v1.35.0): See CHANGE_LOG.md

actions-user committed:
Merge remote-tracking branch 'origin/production'

tomwisecodes committed:
feat: am styling (#655)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Joe Baker <joe.baker@thenational.academy>
Co-authored-by: Stef Lewandowski <stef@stef.io>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mantagen committed:
fix: prisma schema drift [AI-1111] (#663)

gclomax committed:
chore: Update quiz generator settings and improve test coverage

- Change `quizReranker` to "schema-reranker" for consistency in quiz generation.
- Update `quizGenerators` to include both "basedOnRag" and "rag" for enhanced functionality.
- Add a new test case to validate the use of schema reranker with override functionality when no `basedOn` lesson is present.
- Improve logging messages for clarity during quiz generation.

gclomax committed:
chore: Update log level in browser logger for improved debugging

- Comment out previous log level configuration to enhance clarity.
- Set log level to "debug" for better visibility during development and troubleshooting.

gclomax committed:
Merge branch 'main' into feat/rag-quiz-cherrypick

gclomax committed:
wip: adding tests to narrow down bug

gclomax committed:
chore: Update quiz fixtures and generator logic

- Added "Milkman" and "Postman" to prior knowledge and key learning points in CircleTheoremLesson.
- Updated quiz generator to fetch lesson slug from `ragLessonPlan` instead of `lessonPlan`.
- Adjusted fetchExperimentalPatches to set the flag for starter and exit quizzes to false for improved functionality.

mantagen committed:
ensure relevant lessons are passed to quiz builder

stefl committed:
fix: clean up typos and linting issues in codebase (#631)

Co-authored-by: Adam Howard <91115+codeincontext@users.noreply.github.com>
Co-authored-by: mantagen <contact.mantagen@gmail.com>

JBR90 committed:
feat: additional material exports - AI-916 (#654)

stefl committed:
feat: prettier editor config for vscode (#609)

Co-authored-by: mantagen <contact.mantagen@gmail.com>

gclomax committed:
chore: switching to return-first and rag

tomwisecodes committed:
chore: remove temp banner (#669)

mikeritson-oak committed:
Merge pull request #670 from oaknational/rc-20250508-1

build: release candidate

semantic-release-bot committed:
build(release v1.36.0): See CHANGE_LOG.md

mantagen committed:
chore: update openai [AI-1117] (#664)

JBR90 committed:
feat: additional resources - threat detection and moderation - AI-1061 AI-1013 AI-860 [MIGRATION] (#657)

gclomax committed:
fix: changed cohere reranker to rerank-v3.5 due to deprecation of old model endpoint

mantagen committed:
feat: add components for all quiz types (#652)

mantagen committed:
chore: patch dependency updates (#666)

mantagen committed:
fix: lakera persistence issue (#656)

tomwisecodes committed:
feat: styling to match designs (#665)

JBR90 committed:
feat: additional materials - connect frontend to exports - AI-1082 (#667)

gclomax committed:
chore: changing to ml reranker

gclomax committed:
chore: changing recommendation method

gclomax committed:
chore: changing generators

tomwisecodes committed:
chore: important tags (#675)

tomwisecodes committed:
chore: important tags (#675)

gclomax committed:
Merge branch 'main' into feat/rag-quiz-cherrypick

mikeritson-oak reviewed (changes requested):
This PR was reviewed last week, with help from @gclomax, and a number of minor issues found that will need some rework. Details have been provided in Slack and directly to devs. 👍

tomwisecodes committed:
feat: am add exit quiz (#673)

Co-authored-by: Joe Baker <joe.baker@thenational.academy>

mikeritson-oak committed:
Merge pull request #674 from oaknational/rc-20250515-1

build: release candidate

semantic-release-bot committed:
build(release v1.37.0): See CHANGE_LOG.md

actions-user committed:
Merge remote-tracking branch 'origin/production'

mantagen committed:
feat: quasi agentic first iteration [AI-1109] (#671)

mantagen committed:
feat: admin user safety violations [AI-1113] (#662)

JBR90 committed:
feat: connect additional materials moderation and threat to modals and zustand (#668)

Co-authored-by: Tom Wise <79859203+tomwisecodes@users.noreply.github.com>

mikeritson-oak committed:
Merge pull request #682 from oaknational/rc-20250522-1

build: release candidate

semantic-release-bot committed:
build(release v1.38.0): See CHANGE_LOG.md

gclomax committed:
Merge branch 'main' into feat/rag-quiz-cherrypick
